File: 3-LongestSubstringWithoutRepeatingCharacters/3-LongestSubstringWithoutRepeatingCharacters.py
# Last updated: 07/08/2025, 12:51:42
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        longest_length = 0
        current_substring = ''
        start_index = 0
        s_length = len(s)

        i = 0
        while i < s_length:
            c = s[i]
            if c in current_substring:
                if (current_length:=len(current_substring)) > longest_length:
                    longest_length = current_length
                
                i = start_index + current_substring.find(c) + 1
                current_substring = s[i]
                start_index = i
            else:
                current_substring += c
            
            i += 1
        
        if (current_length:=len(current_substring)) > longest_length:
            longest_length = current_length

        return longest_length

File: 3-LongestSubstringWithoutRepeatingCharacters/3-LongestSubstringWithoutRepeatingCharacters.cpp
// Last updated: 07/08/2025, 12:51:38
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> current_substring = {};
        int answer = 0;
        int l = 0; int r = 0;
        for (char c : s) {
            while (current_substring.contains(c)) {
                current_substring.erase(s[l++]);
            }
            current_substring.insert(c);
            r++;
            answer = max(answer, r-l);
        }
        return answer;
    }
};

static const auto _____ = []()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    return nullptr;
}();